walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,127 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {   
+    // Make optional attendees mandatory if there are no mandatory attendees.
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    if (attendees.size() == 0 && optionalAttendees.size() > 0) {
+      attendees = optionalAttendees;
+      optionalAttendees = new ArrayList<String>();
+    }
+
+    if (optionalAttendees.size() == 0) {
+      return FindMeetingTimes(events, request, attendees);
+    }
+    Collection<String> allAttendees = merge(attendees, optionalAttendees);
+    Collection<TimeRange> meetingTimes = FindMeetingTimes(events, request, allAttendees);
+    if (meetingTimes.size() == 0) {
+      return FindMeetingTimes(events, request, attendees);
+    }
```

Comment:
[Optional] It might be worth adding some comments here to clarify the logic.  I didn't realize at first that the desired behavior was that you can't make the meeting worth with all attendees you just ignore the optional ones entirely.